File: Road traffic data/C/hash_table.c
//NOM: CHEYCHOU MOUAFO JUNIOR   mat: 21T2374
#include <stdio.h>
#include <stdlib.h>

#define MAX 300
#define HASHTABLE_SIZE 10

// La structure de donnée qui vas contenir horodateur
typedef struct {
    int jour;
    int mois;
    int ans;
    int heure;
    int min;
    int sec;
}horodate;
// La structure de donnée pour la date
typedef struct {
    int jour;
    int mois;
    int ans;
}date;

typedef struct{
    horodate horodateur;
    char sexe;
    int age;
    date Date;
    char period[10];
    char repas[MAX];
    char unite[30];
    char lieu[25];
    int nombre;
}data;

typedef struct node{
    data value;
    struct node* next;
}node;

void insert_new_node(node** head, data new_value){
    node* new_node = (node*)malloc(sizeof(node));
    new_node->value = new_value;
    new_node->next = NULL;

    if(*head == NULL){
        *head = new_node;
    }else{
        node* current = *head;
        while(current->next!=NULL){
            current = current->next;
        }
        current->next = new_node;
    }
}

int get_location(int new_value){
    return new_value%HASHTABLE_SIZE;
}

void insert_into_table(node** hashtable, data new_value){
    int location = get_location(new_value.age);

    if(hashtable[location]==NULL){
        hashtable[location] = malloc(sizeof(node*));

        node* head = NULL;
        insert_new_node(&head, new_value);

        hashtable[location] = head;
    }else{
        node* head = hashtable[location];
        insert_new_node(&head,new_value);

        hashtable[location] = head;
    }
}

void afficher(node** hashtable){
    int i;
    for(i=0; i< HASHTABLE_SIZE; i++){
        node* head = hashtable[i];

        if(head != NULL){
            node* current = head;
            while(current!=NULL){
                printf("%d/%d/%-5d%d:%d:%-5d",current->value.horodateur.jour,current->value.horodateur.mois,current->value.horodateur.ans,current->value.horodateur.heure,current->value.horodateur.min,current->value.horodateur.sec);
                printf("%-3c%-3d%d/%d/%-7d",current->value.sexe,current->value.age,current->value.Date.jour,current->value.Date.mois,current->value.Date.ans);
                printf("%-10s%-65s%-20s%-15s%-5d\n\n",current->value.period,current->value.repas,current->value.unite,current->value.lieu,current->value.nombre);
                current = current->next;
            }
        }
    }
}

node** get_hashtable(){
    node** hashtable = calloc(HASHTABLE_SIZE, sizeof(node*));
    return hashtable;
}

node **save(){
    node** table = get_hashtable();
    FILE *in = fopen("foodData.txt","r");
    int i;
    data x;
    char ch;

     if(in == NULL){
        printf("Erreur lors de l'ouverture du fichier !!! :(\n");
        return NULL;
     }
     fscanf(in,"%d",&x.horodateur.jour);
     if(feof(in))
        printf("Desolé, le fichier est vide :(\n");
     while(x.horodateur.jour != EOF){
        i = 0;
        if(feof(in))
            break;
        fscanf(in,"%d%d%d%d%d",&x.horodateur.mois,&x.horodateur.ans,&x.horodateur.heure,&x.horodateur.min,&x.horodateur.sec);
        x.sexe = getc(in);//recupere l'espace qu'il ya entre le champs seconde et sexe
        x.sexe = getc(in);//ecrase la valeur precedente(l'espace) avec celui du sexe
        fscanf(in,"%d%d%d%d%s",&x.age,&x.Date.jour,&x.Date.mois,&x.Date.ans,x.period);
        ch = getc(in);
        ch = getc(in);
        while((ch = getc(in))!='"'){
            x.repas[i++]=ch;
        }
        x.repas[i]='\0';//indique la fin de la chaine de caractère
        ch = getc(in);
        ch = getc(in);
        i = 0;
        while((ch = getc(in))!='"'){
            x.unite[i++]=ch;
        }
        x.unite[i]='\0';//indique la fin de la chaine de caractère
        fscanf(in,"%s%d",x.lieu,&x.nombre);
        insert_into_table(table,x);
        fscanf(in,"%d",&x.horodateur.jour);
    }
    fclose(in);
    return table;
}
void search(node** table, int age){
    int index = get_location(age);
    node* head = table[index];
    int i = 0;

    while(head!=NULL){
        if(head->value.age == age){
            printf("%d/%d/%-5d%d:%d:%-5d",head->value.horodateur.jour,head->value.horodateur.mois,head->value.horodateur.ans,head->value.horodateur.heure,head->value.horodateur.min,head->value.horodateur.sec);
            printf("%-3c%-3d%d/%d/%-7d",head->value.sexe,head->value.age,head->value.Date.jour,head->value.Date.mois,head->value.Date.ans);
            printf("%-10s%-65s%-20s%-15s%-5d\n\n",head->value.period,head->value.repas,head->value.unite,head->value.lieu,head->value.nombre);
            i++;
        }
        head = head->next;
    }
    if(i==0)
        printf("Aucun element trouver\n");
}

void sort(node** table){
    int i;
    for(i=7; i< HASHTABLE_SIZE; i++){
        node* head = table[i];

        if(head != NULL){
            node* current = head;
            while(current!=NULL){
                printf("%d/%d/%-5d%d:%d:%-5d",current->value.horodateur.jour,current->value.horodateur.mois,current->value.horodateur.ans,current->value.horodateur.heure,current->value.horodateur.min,current->value.horodateur.sec);
                printf("%-3c%-3d%d/%d/%-7d",current->value.sexe,current->value.age,current->value.Date.jour,current->value.Date.mois,current->value.Date.ans);
                printf("%-10s%-65s%-20s%-15s%-5d\n\n",current->value.period,current->value.repas,current->value.unite,current->value.lieu,current->value.nombre);
                current = current->next;
            }
        }
    }
    for(i=0; i< 7; i++){
        node* head = table[i];

        if(head != NULL){
            node* current = head;
            while(current!=NULL){
                printf("%d/%d/%-5d%d:%d:%-5d",current->value.horodateur.jour,current->value.horodateur.mois,current->value.horodateur.ans,current->value.horodateur.heure,current->value.horodateur.min,current->value.horodateur.sec);
                printf("%-3c%-3d%d/%d/%-7d",current->value.sexe,current->value.age,current->value.Date.jour,current->value.Date.mois,current->value.Date.ans);
                printf("%-10s%-65s%-20s%-15s%-5d\n\n",current->value.period,current->value.repas,current->value.unite,current->value.lieu,current->value.nombre);
                current = current->next;
            }
        }
    }
}

int main(){
    node** table = get_hashtable();
    int choix,age;
    data *tab;

    do{
        printf("---------------------- MENU ------------------------\n");
        printf("0- Quiter\n");
        printf("1- Enregistrer et afficher le tableau\n");
        printf("2- Rechercher et afficher les elements du tableau\n");
        printf("3- Trier le tableau (en fonction de l'age)\n\n");
        printf("Faites un choix : ");
        scanf("%d",&choix);
        switch(choix){
            case 0: break;
            case 1: system("cls");
                    table = save();
                    afficher(table);
                    break;
            case 2: system("cls");
                    printf("Entrer l'age a rechercher : ");
                    scanf("%d",&age);
                    system("cls");
                    search(table,age);
                    break;
            case 3: system("cls");
                    sort(table);
                    printf("Le tri a ete effectuer avec succes :)\n");
                    break;
            default: printf("Faire un choix entre 0 et 3...\n");
        }
        system("PAUSE");
        system("cls");
    }while(choix!=0);
    return 0;
}

File: Road traffic data/Js/hash_table.js
//NOM: CHEYCHOU MOUAFO JUNIOR   mat: 21T2374
function save_table(){
	var line = ["23/03/2022","16:26:22","M",21,"23/03/2022","12h-14h","Riz sauce d'arachide","Plateau","Maison",8];
	var lines = [];
	lines[0] = line;
	line = ["23/03/2022","16:27:24","F",18,"23/03/2022","14h-16h","Corn flakes","Bol","Maison",1];
	lines[lines.length] = line;
	line = ["23/03/2022","16:27:47","M",18,"12/01/2004","14h-16h","Riz suave arachide avec la viande","Plat","Maison",3];
	lines[lines.length] = line;
	line = ["23/03/2022","16:29:33","F",18,"23/03/2022","12h-14h","Purée d'avocat-pain","Plat","Maison",3];
	lines[lines.length] = line;
	line = ["23/03/2022","16:31:02","M",17,"23/03/2022","14h-16h","Rice and tomato stew","Plat","Maison",6];
	lines[lines.length] = line;
	line = ["23/03/2022","16:32:32","M",19,"23/03/2022","14h-16h","Riz sauce blanche","Plat","Maison",6];
	lines[lines.length] = line;
	line = ["23/03/2022","16:32:53","M",18,"23/03/2022","14h-16h","Riz blanc accompagné du haricot","Plat","Maison",7];
	lines[lines.length] = line;
	line = ["23/03/2022","16:33:37","M",22,"23/03/2022","12h-14h","Foufou ans Eru","Plat","Maison",3];
	lines[lines.length] = line;
	line = ["23/03/2022","16:33:52","M",22,"23/03/2022","12h-14h","Beignets haricots Bouilli","plat","Maison",2];
	lines[lines.length] = line;
	line = ["23/03/2022","16:34:02","M",19,"23/03/2022","12h-14h","Plantain frit et 2 oeufs","Assiette creuse","Maison",6];
	lines[lines.length] = line;
	line = ["23/03/2022","16:37:53","F",20,"23/03/2022","16h-18h","Les crêpes au lait","Bol","Maison",6];
	lines[lines.length] = line;
	line = ["23/03/2022","16:40:05","M",21,"14/04/2000","16h-18h","okok baton de manioc","Assiette creuse","Maison",5];
	lines[lines.length] = line;
	line = ["23/03/2022","16:41:40","M",18,"23/03/2022","10h-12h","Pain et une tasse de café au lait","Tasse de café","Maison",1];
	lines[lines.length] = line;
	line = ["23/03/2022","16:42:03","M",20,"22/03/2022","12h-14h","Le riz avec la sauce tomate","Plat","Maison",1];
	lines[lines.length] = line;
	line = ["23/03/2022","16:43:35","F",20,"23/03/2022","20h-22h","Sauce d'arrahide avec le manioc","Plat","Maison",6];
	lines[lines.length] = line;
	line = ["23/03/2022","16:43:40","F",17,"23/03/2022","16h-18h","Riz sauce d'arachide sans viande","Plat","Maison",5];
	lines[lines.length] = line;
	line = ["23/03/2022","16:43:53","F",20,"23/03/2022","14h-16h","Okok sucré ,manioc","Assiette creuse","Maison",6];
	lines[lines.length] = line;
	line = ["23/03/2022","16:45:54","M",36,"23/03/2022","12h-14h","Bouillon viande, plaintain mur","Assiette creuse","Restaurant",0];
	lines[lines.length] = line;
	line = ["23/03/2022","16:46:26","M",18,"23/03/2022","6h-10h","pain omelette","Plat","Route",0];
	lines[lines.length] = line;	
	
	return lines;
}

class hash_table{
    constructor(){
        this.table = [[],[],[],[],[],[],[],[],[],[]]
		this.nbr_elt = [0,0,0,0,0,0,0,0,0,0,];
    }    
    save(){
		var lines = save_table();

		var i = 0;
		while(i < 19){
			this.insert(lines[i]);
			i++;
		} 
	}
        
    insert(val){
        var index = this.hash_function(val[3]);
        if(this.nbr_elt[index] == 0){
            this.table[index][0] = val;
			this.nbr_elt[index]++;
        }else {
            this.table[index][this.nbr_elt[index]] = val;
			this.nbr_elt[index]++;
		}
	}
        
    hash_function(val){
        return val % 10;
    }
	
    afficher(){
        var j,i = 0;
        while(i < 10){
            j = 0;
            while(j < this.table[i].length){
                console.log(this.table[i][j][0]+'\t'+this.table[i][j][1]+'\t'+this.table[i][j][2]+'\t'+this.table[i][j][3]+'\t'+this.table[i][j][4]+'\t'+this.table[i][j][5]+'\t'+this.table[i][j][6]+'\t'+this.table[i][j][7]+'\t'+this.table[i][j][8]+'\t'+this.table[i][j][9]+'\n');
                j++;
			}
            i++;
		}
	}
            
    search(age){
        var index = this.hash_function(age); 
        var j = 0, i = 0;
        if (this.nbr_elt[index] == 0)
            console.log('Aucun element trouver');
        while(i < this.table[index].length){
			if(age == this.table[index][i][3]){
				j++
				console.log(this.table[index][i][0]+'\t'+this.table[index][i][1]+'\t'+this.table[index][i][2]+'\t'+this.table[index][i][3]+'\t'+this.table[index][i][4]+'\t'+this.table[index][i][5]+'\t'+this.table[index][i][6]+'\t'+this.table[index][i][7]+'\t'+this.table[index][i][8]+'\t'+this.table[index][i][9]+'\n');
            }
			i++;
		}
		if(j == 0)
			console.log('Aucun element trouver');
	}
            
    sort(){
        var j,i = 7;
        while(i < 10){
            j = 0;
            while(j < this.nbr_elt[i]){
                console.log(this.table[i][j][0]+'\t'+this.table[i][j][1]+'\t'+this.table[i][j][2]+'\t'+this.table[i][j][3]+'\t'+this.table[i][j][4]+'\t'+this.table[i][j][5]+'\t'+this.table[i][j][6]+'\t'+this.table[i][j][7]+'\t'+this.table[i][j][8]+'\t'+this.table[i][j][9]+'\n');
                j++;
			}
            i++;
		}
        i = 0;
        while(i < 7){
            j = 0;
            while(j < this.nbr_elt[i]){
                console.log(this.table[i][j][0]+'\t'+this.table[i][j][1]+'\t'+this.table[i][j][2]+'\t'+this.table[i][j][3]+'\t'+this.table[i][j][4]+'\t'+this.table[i][j][5]+'\t'+this.table[i][j][6]+'\t'+this.table[i][j][7]+'\t'+this.table[i][j][8]+'\t'+this.table[i][j][9]+'\n');
                j++;
			}
            i++;
		}
	}
}
            
var x = 1
hashe = new hash_table();
while (x != 0){
    console.log("----------------------- MENU --------------------------");
    console.log("0 - Quitter");
    console.log("1 - Enregistrer les donnee");
    console.log("2 - Trier en fonction de l'age");
    console.log("3 - Rechercher les donnee en fonction de l'age\n");
    x = prompt("Faites un choix : ");
    x = parseInt(x);
    if (x == 1){
        hashe.save();
		hashe.afficher();
    }else if(x == 2){
        hashe.sort();
    }else if(x == 3){
        age = prompt("Entrer l'age a rechercher : ");
        age = parseInt(age);
        hashe.search(age);
	}else if(x == 0){
		
    }else
        console.log("Faire un choix entre 0 et 3");
}